icu4c/source/i18n/dtfmtsym.cpp

```diff
@@ -1950,6 +1909,51 @@ initField(UnicodeString **field, int32_t& length, CalendarDataSink &sink, CharSt
     }
 }
 
+static void
+initEras(UnicodeString **field, int32_t& length, CalendarDataSink &sink, CharString &key, const UResourceBundle *ctebPtr, const char* eraWidth, int32_t maxEra, UErrorCode &status) {
+    if (U_SUCCESS(status)) {
+        length = 0;
+        UnicodeString keyUString(key.data(), -1, US_INV);
+        Hashtable *eraNamesTable = static_cast<Hashtable*>(sink.maps.get(keyUString));
+
+        if (eraNamesTable != nullptr) {
+            UErrorCode resStatus = U_ZERO_ERROR;
+            LocalUResourceBundlePointer ctewb(ures_getByKeyWithFallback(ctebPtr, eraWidth, nullptr, &resStatus));
+            const UResourceBundle *ctewbPtr = (U_SUCCESS(resStatus))? ctewb.getAlias() : nullptr;
+            *field = new UnicodeString[maxEra + 1];
+            if (*field == nullptr) {
+                status = U_MEMORY_ALLOCATION_ERROR;
+                return;
+            }
+            length = maxEra + 1;
+            for (int32_t eraCode = 0; eraCode <= maxEra; eraCode++) {
+                char eraCodeStr[12]; // T_CString_integerToString is documented to generate at most 12 bytes including nul terminator
+                int32_t eraCodeStrLen = T_CString_integerToString(eraCodeStr, eraCode, 10);
+                UnicodeString eraCodeKey = UnicodeString(eraCodeStr, eraCodeStrLen, US_INV);
+                UnicodeString *eraName = static_cast<UnicodeString*>(eraNamesTable->get(eraCodeKey));
+                (*field)[eraCode].remove();
+                if (eraName != nullptr) {
+                    // Get eraName from map (created by CalendarSink)
+                    (*field)[eraCode].fastCopyFrom(*eraName);
+                } else if (ctewbPtr != nullptr) {
+                    // Try filling in missing items from parent locale(s)
+                    resStatus = U_ZERO_ERROR;
+                    LocalUResourceBundlePointer ctewkb(ures_getByKeyWithFallback(ctewbPtr, eraCodeStr, nullptr, &resStatus));
+                    if (U_SUCCESS(resStatus)) {
+                        int32_t eraNameLen;
+                        const UChar* eraNamePtr = ures_getString(ctewkb.getAlias(), &eraNameLen, &resStatus);
+                        if (U_SUCCESS(resStatus)) {
+                            (*field)[eraCode].setTo(false, eraNamePtr, eraNameLen);
+                        }
+                    }
+                }
+            }
+            return;
+        }
+        status = U_MISSING_RESOURCE_ERROR;
+    }
+}
+
```

Comment:
The following ticket has been filed to rework the ICU4J code in a similar way: [ICU-23197](https://unicode-org.atlassian.net/browse/ICU-23197)